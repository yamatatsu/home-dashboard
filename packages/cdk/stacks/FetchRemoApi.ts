import * as cdk from "@aws-cdk/core";
import * as lambda from "@aws-cdk/aws-lambda";
import * as events from "@aws-cdk/aws-events";
import * as eventsTargets from "@aws-cdk/aws-events-targets";
import * as sns from "@aws-cdk/aws-sns";
import * as sqs from "@aws-cdk/aws-sqs";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import { SqsSubscription } from "@aws-cdk/aws-sns-subscriptions";
import { SqsEventSource } from "@aws-cdk/aws-lambda-event-sources";

type FetchRemoApiProps = cdk.StackProps & {
  code: lambda.Code;
  remoToken: string;
  homeDataTable: dynamodb.ITable;
};

/**
 * Schedule -> Lambda -> SNS -> SQS -> Lambda -> DynamoDB
 * use SNS for extensibility and my learning ðŸ˜œ
 * use SQS for rerunability and my learning ðŸ˜œ
 */
export class FetchRemoApi extends cdk.Stack {
  public readonly topic: sns.ITopic;

  constructor(scope: cdk.Construct, id: string, props: FetchRemoApiProps) {
    super(scope, id, props);

    const { code, remoToken, homeDataTable } = props;

    const topic = new sns.Topic(this, "Topic");

    const fetchRemoApi = new lambda.Function(this, "fetchRemoApi", {
      functionName: "FetchRemoApi",
      code,
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: "index.fetchRemoApi",
      environment: {
        REMO_TOKEN: remoToken,
        TOPIC_ARN: topic.topicArn,
      },
      memorySize: 128,
      timeout: cdk.Duration.seconds(3),
    });
    const putRemoData = new lambda.Function(this, "putRemoData", {
      functionName: "PutRemoData",
      code,
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: "index.putRemoData",
      environment: {
        TABLE_NAME: homeDataTable.tableName,
      },
      memorySize: 128,
      timeout: cdk.Duration.seconds(3),
    });

    new events.Rule(this, "ScheduleRule", {
      schedule: events.Schedule.rate(cdk.Duration.minutes(10)),
      targets: [new eventsTargets.LambdaFunction(fetchRemoApi)],
    });

    topic.grantPublish(fetchRemoApi);

    const putRemoDataQueue = new sqs.Queue(this, "putRemoDataQueue", {});
    topic.addSubscription(new SqsSubscription(putRemoDataQueue));
    putRemoData.addEventSource(new SqsEventSource(putRemoDataQueue));

    homeDataTable.grantWriteData(putRemoData);
  }
}
