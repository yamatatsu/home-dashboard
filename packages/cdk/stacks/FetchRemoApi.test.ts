import { SynthUtils } from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";
import * as lambda from "@aws-cdk/aws-lambda";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import { FetchRemoApi } from "./FetchRemoApi";

test("FetchRemoApi Stack", () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "test-stack");

  const table = new dynamodb.Table(stack, "test-table", {
    partitionKey: { name: "hoge", type: dynamodb.AttributeType.STRING },
  });

  const target = new FetchRemoApi(app, "MyTestStack", {
    code: lambda.Code.fromInline("xxx"),
    remoToken: "xxx",
    remoRawEventsTable: table,
  });

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
