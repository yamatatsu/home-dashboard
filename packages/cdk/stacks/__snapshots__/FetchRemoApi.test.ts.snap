// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchRemoApi Stack 1`] = `
Object {
  "Resources": Object {
    "ScheduleRuleAllowEventRuleTargetScheduleRule0C75E08003FCA383": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "fetchRemoApi3831C897",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "ScheduleRuleDA5BD877",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ScheduleRuleDA5BD877": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(10 minutes)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "fetchRemoApi3831C897",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TopicBFC7AF6E": Object {
      "Type": "AWS::SNS::Topic",
    },
    "fetchRemoApi3831C897": Object {
      "DependsOn": Array [
        "fetchRemoApiServiceRoleDefaultPolicy0A2B9AC4",
        "fetchRemoApiServiceRole01CA419F",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "xxx",
        },
        "Environment": Object {
          "Variables": Object {
            "REMO_TOKEN": "xxx",
            "TOPIC_ARN": Object {
              "Ref": "TopicBFC7AF6E",
            },
          },
        },
        "FunctionName": "FetchRemoApi",
        "Handler": "index.fetchRemoApi",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "fetchRemoApiServiceRole01CA419F",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 3,
      },
      "Type": "AWS::Lambda::Function",
    },
    "fetchRemoApiServiceRole01CA419F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "fetchRemoApiServiceRoleDefaultPolicy0A2B9AC4": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TopicBFC7AF6E",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "fetchRemoApiServiceRoleDefaultPolicy0A2B9AC4",
        "Roles": Array [
          Object {
            "Ref": "fetchRemoApiServiceRole01CA419F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "putRemoData73055A1E": Object {
      "DependsOn": Array [
        "putRemoDataServiceRoleDefaultPolicy147EFD9F",
        "putRemoDataServiceRoleB14A0DAE",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "xxx",
        },
        "Environment": Object {
          "Variables": Object {
            "TABLE_NAME": Object {
              "Fn::ImportValue": "test-stack:ExportsOutputReftesttableB6F02C7740211428",
            },
          },
        },
        "FunctionName": "PutRemoData",
        "Handler": "index.putRemoData",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "putRemoDataServiceRoleB14A0DAE",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 3,
      },
      "Type": "AWS::Lambda::Function",
    },
    "putRemoDataQueueCC2E67D5": Object {
      "Type": "AWS::SQS::Queue",
    },
    "putRemoDataQueuePolicyCEB7E013": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sqs:SendMessage",
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:SourceArn": Object {
                    "Ref": "TopicBFC7AF6E",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "sns.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "putRemoDataQueueCC2E67D5",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": Array [
          Object {
            "Ref": "putRemoDataQueueCC2E67D5",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "putRemoDataQueueTargetTopicAD1B2EBDBF7F1442": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "putRemoDataQueueCC2E67D5",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": Object {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "putRemoDataServiceRoleB14A0DAE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "putRemoDataServiceRoleDefaultPolicy147EFD9F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "putRemoDataQueueCC2E67D5",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "test-stack:ExportsOutputFnGetAtttesttableB6F02C77Arn4AC37B77",
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "putRemoDataServiceRoleDefaultPolicy147EFD9F",
        "Roles": Array [
          Object {
            "Ref": "putRemoDataServiceRoleB14A0DAE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "putRemoDataSqsEventSourceTargetputRemoDataQueue75D9F93D513CE4DF": Object {
      "Properties": Object {
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "putRemoDataQueueCC2E67D5",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "putRemoData73055A1E",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
  },
}
`;
